constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"fmt"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+var trueBool = true
+
+// AddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// An object can have values not described in the schema.  A blank Type could be
+	// anything, including an object, so we should treat it the same as those types.
+	// typeWithFields := sch.Type == "" || sch.Type == "object"
+	typeWithFields := sch.Type == "" || sch.Type == "object"
+
+	// Don't override an already set XPreserveUnknownFields
+	if typeWithFields && sch.XPreserveUnknownFields == nil {
+		sch.XPreserveUnknownFields = &trueBool
+	}
+
+	if sch.Properties != nil {
+		for k, v := range sch.Properties {
+			if err := AddPreserveUnknownFields(&v); err != nil {
+				return err
+			}
+
+			// As v is not a pointer, we need to set the contents of v back into the original data structure
+			sch.Properties[k] = v
+		}
+	}
+
+	// If the type is array, the schema of the array's items must be structural.  If the schema
+	// is undefined, we must add a blank one with x-kubernetes-preserve-unknown-fields to meet
+	// this structural item schema requirement.
+	schemaInvalid := sch.Items == nil || (sch.Items.Schema == nil && sch.Items.JSONSchemas == nil)
+	if sch.Type == "array" && schemaInvalid {
+		sch.Items = &apiextensionsv1beta1.JSONSchemaPropsOrArray{
+			Schema: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+			},
+		}
+
+		return nil
+	}
+
+	if sch.Items != nil {
+		if sch.Items.Schema != nil {
+			if err := AddPreserveUnknownFields(sch.Items.Schema); err != nil {
+				return err
+			}
+		}
+
+		if sch.Items.JSONSchemas != nil {
+			return fmt.Errorf("non-nil JSONSchemas encountered, multiple schemas are not supported")
+		}
+	}
+
+	if sch.AdditionalProperties != nil && sch.AdditionalProperties.Schema != nil {
```

Comment:
Is this valid for Arrays, or just Objects?

Reply:
`AdditionalProperties` [is meant to provide some schema validation for fields not defined in `properties`](https://json-schema.org/understanding-json-schema/reference/object.html#index-1:~:text=The%20additionalProperties%20keyword%20is%20used%20to,if%20they%20are%20each%20a%20string%3A).  So it's really just another way of defining an object.

Reply:
Was it previously invalid for the user to specify `.AdditionalProperties` for a non-object type?  If so, this could be moved to the object handling portion.

Reply:
The thing that makes me not like splitting the latter half of this function into a type-based switch statement is that all of the logic needs to be called for `Type: ""` and only parts of it are called for `Type: "object"` and `Type: "array"`.  If I'm to avoid repeating code, then I split the various parts of this logic into separate functions.  These functions would then call back recursively to `AddPreserveUnknownFields()`.  Personally I find that more confusing than what I have now.

Is there something I'm not seeing about how this could be put together?